src/ALE/MOM_remapping.F90

```diff
@@ -493,33 +574,85 @@ subroutine remap_via_sub_cells( n0, h0, u0, ppoly0_E, ppoly0_coefs, n1, h1, meth
-  real, dimension(n0) :: u0_min ! Minimum value of reconstructions in source cell [A]
-  real, dimension(n0) :: u0_max ! Minimum value of reconstructions in source cell [A]
   integer, dimension(n1) :: itgt_start ! Index of first sub-cell within each target cell
   integer, dimension(n1) :: itgt_end ! Index of last sub-cell within each target cell
-  real :: xa, xb ! Non-dimensional position within a source cell (0..1) [nondim]
+  ! For error checking/debugging
+  logical, parameter :: force_bounds_in_target = .true. ! To fix round-off issues
+  real :: u02_err ! Integrated reconstruction error estimates [H A]
+
+  ! Calculate sub-layer thicknesses and indices connecting sub-layers to source and target grids
+  call intersect_src_tgt_grids( n0, h0, n1, h1, h_sub, h0_eff, &
+                                isrc_start, isrc_end, isrc_max, itgt_start, itgt_end, isub_src )
+
+  ! Loop over each sub-cell to calculate average/integral values within each sub-cell.
+  ! Uses: h_sub, h0_eff, isub_src
+  ! Sets: u_sub, uh_sub
+  call remap_src_to_sub_grid(n0, h0, u0, ppoly0_E, ppoly0_coefs, n1, h_sub, &
+                             isrc_start, isrc_end, isrc_max, isub_src, &
+                             method, force_bounds_in_subcell, u_sub, uh_sub, u02_err)
+
+  ! Loop over each target cell summing the integrals from sub-cells within the target cell.
+  ! Uses: itgt_start, itgt_end, h1, h_sub, uh_sub, u_sub
+  ! Sets: u1, uh_err
+  call remap_sub_to_tgt_grid(n0, n1, h1, h_sub, u_sub, uh_sub, itgt_start, itgt_end, &
+                             force_bounds_in_target, u1, uh_err)
+
+  ! Include the error remapping from source to sub-cells in the estimate of total remapping error
+  uh_err = uh_err + u02_err
+
+end subroutine remap_via_sub_cells
+
+!> Returns the intersection of source and targets grids along with and auxiliary lists or indices.
+!!
+!! For source grid with thicknesses h0(1:n0) and target grid with thicknesses  h1(1:n1) the intersection
+!! or "subgrid" has thicknesses h_sub(1:n0+n1+1).
+!! h0 and h1 must have the same units. h_sub will return with the same units as h0 and h1.
+!!
+!! Notes on the algorithm:
+!! Internally, grids are defined by the interfaces (although we describe grids via thicknesses for accuracy).
+!! The intersection or union of two grids is thus defined by the super set of both lists of interfaces.
+!! Because both source and target grids can contain vanished cells, we do not eliminate repeated
+!! interfaces from the union.
+!! That is, the total number of interfaces of the sub-cells is equal to the total numer of interfaces of
+!! the source grid (n0+1) plus the total number of interfaces of the target grid (n1+1), i.e. n0+n1+2.
+!! Whenever target and source interfaces align, then the retention of identical interfaces leads to a
+!! vanished subcell.
+!! The remapping uses a common point of reference to the left (top) so there is always a vanished subcell
```

Comment:
I have never been very comfortable with the use of `left` and `right` to describe the directionality of the remapping.  This implies a lack of rotational symmetry that the `top` and `bottom` do not.  I know that this is a carry-over from the original White and Adcroft paper, but I still do not like it.  Would the extensive revisions in this PR be the chance to correct that nomenclature (and also replace variables like `dx` with `dz`) to more accurately tailor it to the usage within MOM6?